File: 21.合并两个有序链表.cpp
/*
 * @lc app=leetcode.cn id=21 lang=cpp
 *
 * [21] 合并两个有序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// #include <iostream>

// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode* next) : val(x), next(next) {}
// };

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* dummyHead = new ListNode();
        ListNode* cur = dummyHead;
        while(list1 != nullptr && list2 != nullptr){
            if(list1->val < list2->val){
                cur->next = list1;
                list1 = list1->next;
            }
            else{
                cur->next = list2;
                list2 = list2->next;
            }
            cur = cur->next;
        }
        if(list1 != nullptr){
            cur->next = list1;
        }
        if(list2 != nullptr){
            cur->next = list2;
        }
        return dummyHead->next;
    }
};
// @lc code=end

// int main(){
//     ListNode* list1 = new ListNode(1, new ListNode(2, new ListNode(4)));
//     ListNode* list2 = new ListNode(1, new ListNode(3, new ListNode(4)));
//     ListNode* p1 = list1;
//     while(p1 != nullptr){
//         printf("%d->", p1->val);
//         p1 = p1->next;
//     }
//     printf("\n");
//     ListNode* p2 = list2;
//     while(p2 != nullptr){
//         printf("%d->", p2->val);
//         p2 = p2->next;
//     }
//     printf("\n");
//     Solution mySolution;
//     ListNode* res = mySolution.mergeTwoLists(list1, list2);
//     ListNode* cur = res;
//     while(cur != nullptr){
//         printf("%d->", cur->val);
//         cur = cur->next;
//     }
//     return 0;
// }

File: 21.合并两个有序链表.java
/*
 * @lc app=leetcode.cn id=21 lang=java
 *
 * [21] 合并两个有序链表
 */

// @lc code=start

import java.util.Scanner;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    // public static class ListNode{
    //     int val;
    //     ListNode next;
    //     ListNode(){};
    //     ListNode(int x){this.val = x;}
    //     ListNode(int x, ListNode next){this.val = x; this.next = next;}
    // }

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummyHead = new ListNode(-1);
        ListNode l1, l2, p;
        l1 = list1;
        l2 = list2;
        p = dummyHead;
        while(l1 != null && l2 != null){
            if(l1.val < l2.val){
                p.next = l1;
                p = p.next;
                l1 = l1.next;
            }
            else{
                p.next = l2;
                p = p.next;
                l2 = l2.next;
            }
        }
        if(l1 != null){
            p.next = l1;
        }
        if(l2 != null){
            p.next = l2;
        }
        return dummyHead.next;
    }

    // public static void main(String[] args){
    //     Scanner myScanner = new Scanner(System.in);
    //     int n, m;
    //     n = myScanner.nextInt();
    //     m = myScanner.nextInt();
    //     ListNode list1, list2;
    //     list1 = new ListNode();
    //     list2 = new ListNode();
    //     for(int i = 0; i < n; i++){
    //         int temp_int = myScanner.nextInt();
    //         ListNode temp = new ListNode(temp_int);
    //         list1.next = temp;
    //     }
    //     for(int i = 0; i < m; i++){
    //         int temp_int = myScanner.nextInt();
    //         ListNode temp = new ListNode(temp_int);
    //         list2.next = temp;
    //     }
    //     Solution solution = new Solution();
    //     ListNode res = solution.mergeTwoLists(list1.next, list2.next);
    //     ListNode p = res;
    //     System.out.print("[");
    //     while(p != null && p.next != null){
    //         System.out.print(p.val);
    //         System.out.print(",");
    //         p = p.next;
    //     }
    //     System.out.print(p.val);
    //     System.out.print("]");
    //     myScanner.close();
    // }
}
// @lc code=end

